### PR TITLE
Standardizes description markdown conversion in API

### LINKED ISSUE #406: Some site fields are not exposed in the API
The site model only renders a small number of fields meaning it is not possible to include latitude and longitude in the filter projection. 

https://github.com/QutEcoacoustics/baw-server/blob/9e6bca814a6e25fb3473a7871a2fefd63b4b829a/app/models/site.rb#L178

```
filter:
{'filter': {'id': {'in': [1792, 1793, 1794, 1177, 1693, 1192, 1193, 1790, 1791, 1462, 1468, 1476, 1106, 1108, 1109, 1494, 1517, 1522, 1523, 1529, 2558, 2559]}}, 'projection': {'include': ['id', 'name', 'latitude', 'longitude']}}
response:
{'meta': {'status': 400, 'message': 'Bad Request', 'error': {'details': 'Filter parameters were not valid: Name must be in [:id, :name, :description], got latitude', 'info': None}}, 'data': None}
```

We should expose the latitude and longitude and possible other fields as well.

### LINKED ISSUE #425: Project route missing fields
Route: `https://staging.ecosounds.org/projects/1`
Method: `GET`
Missing fields:

- `created_at`
- `updater_id`
- `updated_at`
- `owner_id`
- `access_level` (Potentially part of capabilities route)

### LINKED ISSUE #487: Model description html markup
There are a few models which have the `description` and `descriptionHtml` property. `description_html` should be a html version of the markdown formatted `description` parameter. In some models, this relationship is broken and needs to be fixed.

List of potentially affected models to verify:

- [x] AnalysisJob
- [x] Bookmark
- [x] Dataset
- [x] Project (known working)
- [x] SavedSearch
- [x] Script
- [x] Site (known broken)

### PR DESCRIPTION
**Pull request by atruskie**
# Standardizes description markdown conversion in API

Fixes #487

Also adds a bunch of tests. To ensure the description fields are tested and generated we needed to add each class to our new testing framework. This meant adding API tests for every model afected.

Also fixes #406. Fixes #425.

## Changes

- standard description html/markdown outputs
- taglines now with all descriptions
- updated api tests for all affected models
- updated swagger docs for all affected endpoints
- updated metadata output for projects and sites

## Problems

API (and permission) specs in the new format are still in the process of being written - hence the large amount of churn.

At the time of PR creation there are a few tests failinf

## Visual Changes

N/A

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Remove/Reduce warnings from edited files
- [x] Unit tests have been added for changes
- [ ] Ensure CI build is passing

### CODE REVIEW

File: spec/routing/sites_routing_spec.rb [hunk 1]
```diff
@@ -17,10 +17,18 @@
...
     it { expect(post('sites/filter')).to route_to('sites#filter', format: 'json') }
+
+    # shallow
+    it { expect(get('sites')).to route_to('sites#index', format: 'json') }
+    it { expect(post('sites')).to route_to('sites#create', format: 'json') }
+    it { expect(get('sites/new')).to route_to('sites#new', format: 'json') }
+    it { expect(put('sites/1')).to route_to('sites#update', format: 'json', id: '1') }
+    it { expect(patch('sites/1')).to route_to('sites#update', format: 'json', id: '1') }
+    it { expect(delete('sites/1')).to route_to('sites#destroy', format: 'json', id: '1') }
+    it { expect(post('sites/filter')).to route_to('sites#filter', format: 'json') }
```
Comment (Allcharles) [inline spec/routing/sites_routing_spec.rb:32]: Double up of `'sites/filter'` tests?

**Review by Allcharles — APPROVED** [review]
Looks good. Only concern I have, which can be fixed in the future, is that the unit tests don't seem to be testing edge cases yet. Only the successful cases are being verified.

File: app/modules/api/schema.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        deleted_at: { type: ['null', 'date'], readOnly: true }
+      }
+    end
+
+    def self.creator_ids_and_ats
+      {
+        creator_id: { '$ref' => '#/components/schemas/id', readOnly: true },
+        created_at: { type: 'date', readOnly: true }
+      }
+    end
```
[the remark this answers is not in the document]
Comment (atruskie) [reply app/modules/api/schema.rb:37]: I removed the duplicated definitions and made the compound definitions depend on the primitives

File: app/modules/api/schema.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    def self.rendered_markdown(attr)
+      {
+        "#{attr}": { type: ['string', 'null'] },
+        "#{attr}_html": { type: ['string', 'null'], readOnly: true },
+        "#{attr}_html_tagline": { type: ['string', 'null'], readOnly: true }
+      }
+    end
```
[the remark this answers is not in the document]
Comment (atruskie) [reply app/modules/api/schema.rb:45]: It is only convention that so far we've stored markup in database columns with the name `description`. There could be other fields that need markdown rendered (say a license text blob).

File: app/modules/custom_render.rb [hunk 4]
```diff
@@ -2,30 +2,29 @@
...
-      ).to_html
+      html = Kramdown::Document.new(value, KRAMDOWN_OPTIONS).to_html
 
       if inline
-        ApplicationController.helpers.sanitize(html, tags: ['strong', 'em'])
+        sanitized = ApplicationController
+                    .helpers
+                    .sanitize(html, tags: ['strong', 'em'])
+                    .squish
+        truncated = sanitized.truncate_words(words)
```
[the remark this answers is not in the document]
Comment (atruskie) [reply app/modules/custom_render.rb:25]: The client should still truncate if the text overlaps the available space?

[the remark this answers is not in the document]
Comment (Allcharles) [reply app/modules/custom_render.rb:25]: True, but I thought the purpose of this was to be able to remove that functionality from the client side

File: app/modules/custom_render.rb [hunk 5]
```diff
@@ -2,30 +2,29 @@
...
-        value,
-        { input: 'GFM', hard_wrap: false }
-      ).to_html
+      html = Kramdown::Document.new(value, KRAMDOWN_OPTIONS).to_html
 
       if inline
-        ApplicationController.helpers.sanitize(html, tags: ['strong', 'em'])
+        sanitized = ApplicationController
+                    .helpers
+                    .sanitize(html, tags: ['strong', 'em'])
```
[the remark this answers is not in the document]
Comment (atruskie) [reply app/modules/custom_render.rb:23]: No, they're the only approved tags for inline markdown output. Inline output is text we want to fit nicely into a sentance without any block-level formatting

 See the else case which uses a default set of safe tags (including the ones you suggested).

File: app/modules/custom_render.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (atruskie) [reply app/modules/custom_render.rb:25]: With what we're outputting here client should be able to use simple CSS overflow props. Truncating chars introduces some problems with breaking up tags

File: app/models/project.rb [hunk 6]
```diff
@@ -110,15 +119,33 @@ def self.schema
...
+        :creator_id,
+        :created_at,
+        :updater_id,
+        :updated_at,
+        :deleter_id,
+        :deleted_at,
+        :owner_ids,
+        :site_ids,
+        :image_urls
+      ]
```
[the remark this answers is not in the document]
Comment (atruskie) [reply app/models/project.rb:148]: Yeah, I agree, however, that is at ask for a different time. See https://github.com/QutEcoacoustics/baw-server/issues/480

File: app/models/saved_search.rb [hunk 7]
```diff
@@ -66,6 +66,34 @@ def self.filter_settings
...
+        :id,
+        :name,
+        :description,
+        :description_html,
+        :description_html_tagline,
+        :stored_query,
+        :creator_id,
+        :created_at,
+        :deleter_id,
+        :deleted_at
```
[the remark this answers is not in the document]
Comment (atruskie) [reply app/models/saved_search.rb:92]: Required means the field must be present for the object to be considered valid.

Note, these schemas are only for output of API. So to have a valid API response (from show/index/filter) the following properties must be present in the response

File: app/models/site.rb [hunk 8]
```diff
@@ -239,4 +244,45 @@ def self.filter_settings
...
+        **Api::Schema.rendered_markdown(:description),
+        **Api::Schema.all_ids_and_ats,
+        #notes: { type: 'object' }, # TODO: https://github.com/QutEcoacoustics/baw-server/issues/467
+        notes: { type: 'string' },
+        project_ids: { type: 'array', items: { type: 'integer' } },
+        location_obfuscated: { type: 'boolean' },
+        custom_latitude: { type: ['number', 'null'], minimum: -90, maximum: 90 },
+        custom_longitude: { type: ['number', 'null'], minimum: -180, maximum: 180 },
+        timezone_information: Api::Schema.timezone_information,
+        image_urls: Api::Schema.image_urls
```
[the remark this answers is not in the document]
Comment (atruskie) [reply app/models/site.rb:264]: I can currently find no issue that says I should be outputting those two fields instead of the `custom_*` ones

File: spec/api/analysis_job_spec.rb [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  end
+
+  path '/analysis_jobs/{id}' do
+    with_id_route_parameter
+    let(:id) { analysis_job.id }
+
+    before do
+      # can only delete an analysis job if it is suspended or completed
+      analysis_job.update_column(:overall_status, :completed)
+    end
```
[the remark this answers is not in the document]
Comment (atruskie) [reply spec/api/analysis_job_spec.rb:57]: Morre in depth integration tests can be found here: https://github.com/QutEcoacoustics/baw-server/blob/e9489c554727f27620f20ed3fd3cb0186b8dd0c9/spec/requests/analysis_jobs_spec.rb#L1021-L1027 and elsewhere in that file

File: spec/api/site_spec.rb [hunk 10]
```diff
@@ -0,0 +1,94 @@
+require 'swagger_helper'
```
[the remark this answers is not in the document]
Comment (atruskie) [reply spec/api/site_spec.rb:1]: I suppppppposssseeeeee....

File: app/models/site.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (atruskie) [reply app/models/site.rb:264]: Made an issue: https://github.com/QutEcoacoustics/baw-server/issues/490
